File: MetadataStoreInterface_first_code.cpp

#include <cstdint>
#include <fstream>
#include <map>
#include <mutex>
#include <optional>
#include <stdexcept>
#include <vector>

class MetadataStore {
public:
    static constexpr size_t PAGE_SIZE = 4096;
    static constexpr size_t FIXED_KEY_SIZE = 8;
    static constexpr size_t FIXED_VAL_SIZE = 8;

    MetadataStore(const std::string& filename) {
        _lock = std::make_unique<std::mutex>();
        _free_size = PAGE_SIZE - 8;  // Initial free size (excluding num_items)

        // Open the file in read-write binary mode
        _file.open(filename, std::ios::in | std::ios::out | std::ios::binary);
        if (!_file) {
            throw std::runtime_error("Failed to open file");
        }

        // Check if the file is initialized
        _file.seekg(0, std::ios::end);
        std::streampos file_size = _file.tellg();
        _file.seekg(0, std::ios::beg);

        if (file_size < PAGE_SIZE) {
            // Initialize the file with zeros
            std::vector<uint8_t> zeros(PAGE_SIZE, 0);
            _file.write(reinterpret_cast<char*>(zeros.data()), PAGE_SIZE);
            _file.flush();
        }

        // Read the entire page from disk
        std::vector<uint8_t> page_data(PAGE_SIZE);
        _file.read(reinterpret_cast<char*>(page_data.data()), PAGE_SIZE);

        // Parse the number of items
        uint64_t num_items = 0;
        for (int i = 0; i < 8; ++i) {
            num_items = (num_items << 8) | page_data[i];
        }

        // Parse key-value pairs and populate the in-memory map
        size_t offset = 8;
        for (uint64_t i = 0; i < num_items; ++i) {
            std::vector<uint8_t> key(page_data.begin() + offset, page_data.begin() + offset + FIXED_KEY_SIZE);
            offset += FIXED_KEY_SIZE;
            std::vector<uint8_t> value(page_data.begin() + offset, page_data.begin() + offset + FIXED_VAL_SIZE);
            offset += FIXED_VAL_SIZE;
            _map[key] = value;
        }

        // Calculate free space
        _free_size = PAGE_SIZE - (8 + num_items * (FIXED_KEY_SIZE + FIXED_VAL_SIZE));
    }

    std::optional<std::vector<uint8_t>> get(const std::vector<uint8_t>& key) {
        std::lock_guard<std::mutex> lock(*_lock);
        auto it = _map.find(key);
        if (it != _map.end()) {
            return it->second;
        }
        return std::nullopt;
    }

    void set(const std::vector<uint8_t>& key, const std::vector<uint8_t>& value) {
        if (key.size() != FIXED_KEY_SIZE || value.size() != FIXED_VAL_SIZE) {
            throw std::invalid_argument("Key and value must be of fixed size");
        }

        std::lock_guard<std::mutex> lock(*_lock);
        
        // Check if there's enough free space
        size_t required_space = 0;
        if (_map.find(key) == _map.end()) {
            required_space += FIXED_KEY_SIZE + FIXED_VAL_SIZE;
        }

        if (_free_size < required_space) {
            throw std::runtime_error("Not enough free space");
        }

        // Update the in-memory map
        _map[key] = value;

        // Write the entire page to disk
        _write_page();

        // Update the free space
        if (_map.find(key) == _map.end()) {
            _free_size -= required_space;
        }
    }

    void delete_key(const std::vector<uint8_t>& key) {
        std::lock_guard<std::mutex> lock(*_lock);
        auto it = _map.find(key);
        if (it != _map.end()) {
            _map.erase(it);
            _write_page();
            _free_size += FIXED_KEY_SIZE + FIXED_VAL_SIZE;
        }
    }

    void close() {
        std::lock_guard<std::mutex> lock(*_lock);
        _write_page();
        _file.close();
    }

    size_t free_size() {
        std::lock_guard<std::mutex> lock(*_lock);
        return _free_size;
    }

private:
    void _write_page() {
        // Create a byte buffer of PAGE_SIZE
        std::vector<uint8_t> buffer(PAGE_SIZE, 0);

        // Write the number of items to the buffer
        uint64_t num_items = _map.size();
        for (int i = 7; i >= 0; --i) {
            buffer[i] = num_items & 0xFF;
            num_items >>= 8;
        }

        // Iterate through the in-memory map, writing each key-value pair to the buffer
        size_t offset = 8;
        for (const auto& [key, value] : _map) {
            std::copy(key.begin(), key.end(), buffer.begin() + offset);
            offset += FIXED_KEY_SIZE;
            std::copy(value.begin(), value.end(), buffer.begin() + offset);
            offset += FIXED_VAL_SIZE;
        }

        // Write the entire buffer to disk at offset 0
        _file.seekp(0, std::ios::beg);
        _file.write(reinterpret_cast<char*>(buffer.data()), PAGE_SIZE);

        // Flush to ensure durability
        _file.flush();
    }

    std::unique_ptr<std::mutex> _lock;
    std::fstream _file;
    std::map<std::vector<uint8_t>, std::vector<uint8_t>> _map;
    size_t _free_size;
};

File: MetadataStoreInterface_first_code.rs

use std::collections::HashMap;
use std::fs::{File, OpenOptions};
use std::io::{Read, Write, Seek, SeekFrom};
use std::sync::Mutex;
use byteorder::{BigEndian, ReadBytesExt, WriteBytesExt};

const PAGE_SIZE: usize = 4096;
const FIXED_KEY_SIZE: usize = 8;
const FIXED_VAL_SIZE: usize = 8;

// Define the MetadataStoreInterface trait
pub trait MetadataStoreInterface {
    fn get(&self, key: &[u8]) -> Option<Vec<u8>>;
    fn set(&self, key: &[u8], value: &[u8]) -> Result<(), String>;
    fn delete(&self, key: &[u8]) -> Result<(), String>;
    fn close(&self) -> Result<(), String>;
    fn free_size(&self) -> usize;
}

pub struct MetadataStore {
    file: Mutex<File>,
    map: Mutex<HashMap<Vec<u8>, Vec<u8>>>,
    free_size: Mutex<usize>,
}

impl MetadataStore {
    pub fn new(filename: &str) -> Result<Self, String> {
        let file = OpenOptions::new()
            .read(true)
            .write(true)
            .create(true)
            .open(filename)
            .map_err(|e| format!("Failed to open file: {}", e))?;

        let mut store = MetadataStore {
            file: Mutex::new(file),
            map: Mutex::new(HashMap::new()),
            free_size: Mutex::new(PAGE_SIZE - 8), // Initial free size (excluding num_items)
        };

        store.initialize()?;
        Ok(store)
    }

    fn initialize(&mut self) -> Result<(), String> {
        let mut file = self.file.lock().unwrap();
        let file_size = file.metadata().map_err(|e| format!("Failed to get file metadata: {}", e))?.len();

        if file_size < PAGE_SIZE as u64 {
            // Initialize the file with zeros
            file.set_len(PAGE_SIZE as u64).map_err(|e| format!("Failed to set file length: {}", e))?;
            file.seek(SeekFrom::Start(0)).map_err(|e| format!("Failed to seek: {}", e))?;
            file.write_all(&vec![0; PAGE_SIZE]).map_err(|e| format!("Failed to write zeros: {}", e))?;
        }

        // Read the entire page from disk
        let mut page_data = vec![0; PAGE_SIZE];
        file.seek(SeekFrom::Start(0)).map_err(|e| format!("Failed to seek: {}", e))?;
        file.read_exact(&mut page_data).map_err(|e| format!("Failed to read page: {}", e))?;

        // Parse the number of items
        let num_items = (&page_data[0..8]).read_u64::<BigEndian>().map_err(|e| format!("Failed to read num_items: {}", e))?;

        // Parse key-value pairs and populate the in-memory map
        let mut map = self.map.lock().unwrap();
        let mut offset = 8;
        for _ in 0..num_items {
            let key = page_data[offset..offset + FIXED_KEY_SIZE].to_vec();
            offset += FIXED_KEY_SIZE;
            let value = page_data[offset..offset + FIXED_VAL_SIZE].to_vec();
            offset += FIXED_VAL_SIZE;
            map.insert(key, value);
        }

        // Calculate free space
        let used_space = 8 + (num_items as usize) * (FIXED_KEY_SIZE + FIXED_VAL_SIZE);
        *self.free_size.lock().unwrap() = PAGE_SIZE - used_space;

        Ok(())
    }

    fn write_page(&self) -> Result<(), String> {
        let mut buffer = vec![0; PAGE_SIZE];
        let map = self.map.lock().unwrap();

        // Write the number of items to the buffer
        (&mut buffer[0..8]).write_u64::<BigEndian>(map.len() as u64)
            .map_err(|e| format!("Failed to write num_items: {}", e))?;

        // Iterate through the in-memory map, writing each key-value pair to the buffer
        let mut offset = 8;
        for (key, value) in map.iter() {
            buffer[offset..offset + FIXED_KEY_SIZE].copy_from_slice(key);
            offset += FIXED_KEY_SIZE;
            buffer[offset..offset + FIXED_VAL_SIZE].copy_from_slice(value);
            offset += FIXED_VAL_SIZE;
        }

        // Write the entire buffer to disk at offset 0
        let mut file = self.file.lock().unwrap();
        file.seek(SeekFrom::Start(0)).map_err(|e| format!("Failed to seek: {}", e))?;
        file.write_all(&buffer).map_err(|e| format!("Failed to write page: {}", e))?;
        file.flush().map_err(|e| format!("Failed to flush: {}", e))?;

        Ok(())
    }
}

impl MetadataStoreInterface for MetadataStore {
    fn get(&self, key: &[u8]) -> Option<Vec<u8>> {
        let map = self.map.lock().unwrap();
        map.get(key).cloned()
    }

    fn set(&self, key: &[u8], value: &[u8]) -> Result<(), String> {
        if key.len() != FIXED_KEY_SIZE || value.len() != FIXED_VAL_SIZE {
            return Err("Key and value must be of fixed size".to_string());
        }

        let mut map = self.map.lock().unwrap();
        let mut free_size = self.free_size.lock().unwrap();

        // Check if there's enough free space
        let required_space = if !map.contains_key(key) {
            FIXED_KEY_SIZE + FIXED_VAL_SIZE
        } else {
            0
        };

        if *free_size < required_space {
            return Err("Not enough free space".to_string());
        }

        // Update the in-memory map
        map.insert(key.to_vec(), value.to_vec());

        // Write the entire page to disk
        drop(map); // Release the lock before calling write_page
        self.write_page()?;

        // Update the free space
        *free_size -= required_space;

        Ok(())
    }

    fn delete(&self, key: &[u8]) -> Result<(), String> {
        let mut map = self.map.lock().unwrap();
        let mut free_size = self.free_size.lock().unwrap();

        if map.remove(key).is_some() {
            drop(map); // Release the lock before calling write_page
            self.write_page()?;
            *free_size += FIXED_KEY_SIZE + FIXED_VAL_SIZE;
        }

        Ok(())
    }

    fn close(&self) -> Result<(), String> {
        // Flush any pending writes
        self.write_page()?;
        
        // The file will be automatically closed when the MetadataStore is dropped
        Ok(())
    }

    fn free_size(&self) -> usize {
        *self.free_size.lock().unwrap()
    }
}

// Implement Drop trait to ensure the file is properly closed
impl Drop for MetadataStore {
    fn drop(&mut self) {
        // The file will be automatically closed when the Mutex<File> is dropped
    }
}
